test/envoye2e/driver/scenario.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package driver
+
+import (
+	"bytes"
+	"log"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	"github.com/ghodss/yaml"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+)
+
+type (
+	Params struct {
+		XDS    int
+		Config cache.SnapshotCache
+		Vars   map[string]string
+		N      int
+	}
+	Step interface {
+		Run(*Params) error
+		Cleanup()
+	}
+	Scenario struct {
+		Steps []Step
+	}
+	// Repeat a step either for N number or duration
+	Repeat struct {
+		N        int
+		Duration time.Duration
+		Step     Step
+	}
+	Sleep struct {
+		time.Duration
+	}
+	// Fork will copy params to avoid concurrent access
+	Fork struct {
+		Fore Step
+		Back Step
+	}
+)
+
+var _ Step = &Repeat{}
+
+func (r *Repeat) Run(p *Params) error {
+	if r.Duration != 0 {
+		start := time.Now()
+		p.N = 0
+		for {
+			if time.Since(start) >= r.Duration {
+				break
+			}
+			log.Printf("repeat %d %v out of %v", p.N, time.Since(start), r.Duration)
```

Comment:
nit: for better error : %v elapsed out of %v ?

Reply:
done, thanks.